# Educational Equity in New Jersey:
## Analyzing Relationships Between Resources and Outcomes Across School Districts

-----------------------------------------------------

### Christy Hernandez
#### CS 668 - Pace University - Fall 2024

---------------------------------------------------------------------------------------------------

As said in my Project Overview Statement:

New Jersey is my home state – as well as home to a diverse population and a wide range of school districts, from affluent suburban areas to urban centers and rural communities. However, this diversity presents a challenge: significant disparities in educational resources, funding, and student outcomes persist across different districts. This imbalance in resources has direct implications on student performance, educational opportunities, and overall equity within the state's educational system. Addressing these disparities, especially the most impactful ones, can lead to more targeted policy interventions and an overall improvement in educational outcomes across New Jersey.

In response to this opportunity, the goal of this project is to conduct a comprehensive analysis of educational equity across school districts in New Jersey, identifying relationships between resources and student outcomes. By examining factors such as funding, teacher-student ratios, and access to technology, the analysis will identify which variables have the most impact on student success. The project will evaluate and propose evidence-based interventions, like increased school funding or mentorship programs, aimed at closing the achievement gap. The final deliverable will be a machine learning model that predicts the potential impact of these interventions on educational outcomes, particularly in underserved communities. This model will allow us to highlight specific areas where inequities are most pronounced and propose targeted, data-driven solutions to address them. By leveraging a pre-existing and publicly available dataset from the State of New Jersey, I aim to generate actionable insights that can guide state and local policymakers in promoting a more equitable distribution of resources and improving educational outcomes.

----------------------

# Data Loading

Let's get started by importing pandas and a library that assists with excel files so we can load the related dataset I found!

In [145]:
!pip install openpyxl

In [146]:
import pandas as pd

Uncomment and use this line of code (being sure to update the path to your own) if you are running this in Jupyter Notebook or similar:

In [147]:
#dataset = pd.read_excel('C:/Users/Christy Hernandez/OneDrive/Desktop/Pace Masters Work/CS Analytics Capstone Project/Source files/Database_DistrictStateDetail.xlsx', sheet_name = None)

If you are running this in Google Colab or similar (which is what I am doing at the moment), upload the file and use this line of code:

In [148]:
dataset = pd.read_excel('Database_DistrictStateDetail.xlsx', sheet_name = None)

Luckily, this dataset has a lot of features that should be able to provide a good analysis overall. Let's see just how many features there are:

In [179]:
print(f'There are {len(dataset)} features.')

There are 72 features.


Okay, and what are they?

In [178]:
print(dataset.keys())

dict_keys(['Important 2022-2023 Notes', 'Header and Contact', 'EnrollmentTrendsbyGrade', 'EnrollmentTrendsByStudentGroup', 'EnrollmentByRacialEthnicGroup', 'PreKAndK-FullDayHalfDay', 'EnrollmentTrendsFullSharedTime', 'EnrollmentByHomeLanguage', 'StudentGrowthTrends', 'StudentGrowth', 'StudentGrowthByGrade', 'StudentGrowthByPerformLevel', 'ELAMathPerformanceTrends', 'ELAParticipationPerformance', 'ELAPerformanceTrends', 'ELAPerformanceByGrade', 'MathParticipationPerformance', 'MathPerformanceTrends', 'MathPerformanceByGradeTest', 'ScienceAssessmentSummaryByGrade', 'ScienceAssessmentByGrade', 'AlternateAssessmentParticipatio', 'EnglishLangProgressToProficienc', 'NJGPA', 'EnglishLangParticipationPerform', 'NAEP', 'PSAT-SAT-ACTParticipation', 'PSAT-SAT-ACTPerformance', 'APIBCourseworkPartPerf', 'APIBDualEnrPartByStudentGrp', 'APIBCoursesOffered', 'CTE_SLEParticipation', 'CTEParticipationByStudentGroup', 'WorkbasedLearningByCareerClust', 'IndustryValuedCredentialsEarned', 'MathCoursePartici

Okay, well, 72 is a lot, and I was sure that not all of them would be very useful here. So, I went through the dataset manually and found 9 features that I think will assist with this project the most! Here is what they are and what they each contain:

1.	'EnrollmentTrendsByStudentGroup'
o	Percentages of district student population by gender, economic disadvantage, disability, English learning, homelessness, foster care, military-connection, and migrant

2.	‘PSAT-SAT-ACTParticipation’
o	Percentage of student participation for each standardized exam type for each district, and then the state average participation rate for each exam.

3.	‘APIBCourseworkPartPerf’
o	Percentage of students in each district that are enrolled in one or more AP or IB course, then percentage for one or more AP or IB exam, then percentage for one or more dual enrollment course, followed by the state averages for each.

4.	‘4YrGraduationCohortProfile’
o	After 4 years of enrollment, for each student group within each district, the percentage of students graduating, then the percentage continuing, then the percentage not continuing, followed by the state averages for each.

5.	‘ChronicAbsenteeism'
o	For each student group for each district, the percentage of students who are chronically absent, as well as the districtwide percentage.

6.	'ViolenceVandalismHIBSubstanceOf'
o	For each district, incidents per 100 students enrolled

7.	‘TeachersExperience’
o	For each district, the average number of years of total experience, then the average number of years of experience in their current district, followed by the state averages for both

8.	'StudentToStaffRatios'
o	For each district, staff to student ratios for different staff positions – Teacher, Admin, Librarian, Nurse, Counselor, etc., as well as a Teacher to Admin ratio.

9.	‘TeachersAdminsLevelofEducation’
o	For teachers and then administrators in each district, the percentage that has achieved a maximum of a bachelors, then a maximum of a masters, then a maximum of a doctorate.

Now that we know which features we want to use, let's create a separate dictionary that has only these features.

In [180]:
desired_data = {}

In [181]:
desired_sheets = ['EnrollmentTrendsByStudentGroup', 'PSAT-SAT-ACTParticipation', 'APIBCourseworkPartPerf', '4YrGraduationCohortProfile', 'ChronicAbsenteeism', 'ViolenceVandalismHIBSubstanceOf', 'TeachersExperience', 'StudentToStaffRatios', 'TeachersAdminsLevelOfEducation']

In [182]:
desired_data = {sheet_name: dataset[sheet_name] for sheet_name in desired_sheets if sheet_name in dataset}

Now let's check how many features exist, just to make sure.

In [183]:
print(f'There are {len(desired_data)} features.\n')

There are 9 features.



Great, now let's take a look at our modified dataset!

In [184]:
for sheet_name, sheet_data in desired_data.items():
    print(f'In regards to the "{sheet_name}" sheet:\n')
    print(desired_data[sheet_name].head(),"\n---------------------------------------\n")

In regards to the "EnrollmentTrendsByStudentGroup" sheet:

  CountyCode CountyName DistrictCode  \
0         01   Atlantic         0010   
1         01   Atlantic         0110   
2         01   Atlantic         0120   
3         01   Atlantic         0125   
4         01   Atlantic         0570   

                                       DistrictName Female Male  \
0                   Absecon Public Schools District     47   53   
1                     Atlantic City School District     48   52   
2        Atlantic County Vocational School District     57   42   
3  Atlantic County Special Services School District     28   72   
4                 Brigantine Public School District     52   48   

  Non-Binary/Undesignated Gender  Economically Disadvantaged Students  \
0                            ≤1%                                 49.4   
1                            ≤1%                                 87.1   
2                            ≤1%                                 53.8   
3    

Great! Let's proceed.

# Target Variable

As the target variable for this project, we will be using the data from the 'Graduates' column of the `4YrGraduationCohortProfile` sheet. We also only want to look at data from the overall districtwide graduation rates for each district, not each individual student group, so let's make sure our target dataframe reflects that.

In [193]:
grad_data = desired_data['4YrGraduationCohortProfile'][desired_data['4YrGraduationCohortProfile']['StudentGroup'] == 'Districtwide']
grad_data = grad_data[['DistrictName','Graduates', 'State: Graduates']]

Let's ensure the data is stored as a float and preform a quick statistical analysis on this graduation data.

In [194]:
grad_data['Graduates'] = pd.to_numeric(grad_data['Graduates'], errors='coerce').astype(float)
grad_data['State: Graduates'] = pd.to_numeric(grad_data['State: Graduates'], errors='coerce').astype(float)
grad_data['Graduates'].describe()

count    314.000000
mean      91.771019
std        7.500157
min       34.700000
25%       89.250000
50%       93.750000
75%       96.375000
max      100.000000
Name: Graduates, dtype: float64

In [195]:
grad_data.head()

DistrictName  Graduates  \
0                      Atlantic City School District       79.8   
17        Atlantic County Vocational School District       97.7   
34                    Buena Regional School District       88.2   
51               Egg Harbor Township School District       93.7   
68  Greater Egg Harbor Regional High School District       87.6   

    State: Graduates  
0               91.1  
17              91.1  
34              91.1  
51              91.1  
68              91.1

Okay, now in order to further explore this variable, let's bring in the addresses of each school district and save that geographical data as its own dataframe.

In [196]:
print(dataset['Header and Contact'].columns)

Index(['COUNTY_CODE', 'COUNTY_NAME', 'DISTRICT_CODE', 'DISTRICT_NAME',
       'GRADESPAN', 'SUPERINTENDENT', 'ADDRESS', 'CITY_STATE_ZIP', 'PHONE',
       'EMAIL', 'WEBSITE', 'FACEBOOK', 'TWITTER', 'District Notes'],
      dtype='object')


In [197]:
geo_data = dataset['Header and Contact'][['DISTRICT_NAME','ADDRESS','CITY_STATE_ZIP']]

In [198]:
geo_data.head()

DISTRICT_NAME                   ADDRESS  \
0                   Absecon Public Schools District         800 Irelan Avenue   
1                     Atlantic City School District      1300 Atlantic Avenue   
2        Atlantic County Vocational School District      5080 Atlantic Avenue   
3  Atlantic County Special Services School District    4805 Nawakwa Boulevard   
4                 Brigantine Public School District  301 East Evans Boulevard   

           CITY_STATE_ZIP  
0        Absecon NJ 08201  
1  Atlantic City NJ 08401  
2   Mays Landing NJ 08330  
3   Mays Landing NJ 08330  
4     Brigantine NJ 08203

Let's also set the indexes of both of these dataframes to the district name, to make merging easier.

In [199]:
grad_data.set_index('DistrictName', inplace=True)
geo_data.rename(columns={'DISTRICT_NAME': 'DistrictName'}, inplace=True)
geo_data.set_index('DistrictName', inplace=True)

<ipython-input-199-69a14014db7f>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  geo_data.rename(columns={'DISTRICT_NAME': 'DistrictName'}, inplace=True)


Now we can merge the two dataframes of graduate and geographical data based on the school district name.

In [200]:
merged_data = pd.merge(grad_data, geo_data, left_index=True, right_index=True, how='left')

In order to use the address information effectively, let's make a new column called 'full_address' where we combine the 'ADDRESS' and 'CITY_STATE_ZIP' columns to create a full address.

In [201]:
merged_data['full_address'] = merged_data['ADDRESS'] + ', ' + merged_data['CITY_STATE_ZIP']

Okay, what does it look like?

In [202]:
merged_data.head()

Graduates  State: Graduates  \
DistrictName                                                                    
Atlantic City School District                          79.8              91.1   
Atlantic County Vocational School District             97.7              91.1   
Buena Regional School District                         88.2              91.1   
Egg Harbor Township School District                    93.7              91.1   
Greater Egg Harbor Regional High School District       87.6              91.1   

                                                                      ADDRESS  \
DistrictName                                                                    
Atlantic City School District                            1300 Atlantic Avenue   
Atlantic County Vocational School District               5080 Atlantic Avenue   
Buena Regional School District                                914 Main Avenue   
Egg Harbor Township School District                            13 Swift Drive   
Greater Egg Harbor Regional High School District  1824 Dr. Dennis Foreman Dr.   

                                                                CITY_STATE_ZIP  \
DistrictName                                                                     
Atlantic City School District                           Atlantic City NJ 08401   
Atlantic County Vocational School District               Mays Landing NJ 08330   
Buena Regional School District                               Richland NJ 08350   
Egg Harbor Township School District               Egg Harbor Township NJ 08234   
Greater Egg Harbor Regional High School District    Mays Landing NJ 08330-2640   

                                                                                       full_address  
DistrictName                                                                                         
Atlantic City School District                          1300 Atlantic Avenue, Atlantic City NJ 08401  
Atlantic County Vocational School District              5080 Atlantic Avenue, Mays Landing NJ 08330  
Buena Regional School District                                   914 Main Avenue, Richland NJ 08350  
Egg Harbor Township School District                    13 Swift Drive, Egg Harbor Township NJ 08234  
Greater Egg Harbor Regional High School District  1824 Dr. Dennis Foreman Dr., Mays Landing NJ 0...

Good, but we should also add the country to each address just to be more complete. In this case, all districts are within the United States, so we can just add 'USA' to each one. While we're at it, we can also set the index to be the code of the district, to make merging dataframes easier in the future.

In [203]:
merged_data['full_address'] = merged_data['full_address'] + ", USA"

This should be it! Let's check it out again.

In [204]:
merged_data.head()

Graduates  State: Graduates  \
DistrictName                                                                    
Atlantic City School District                          79.8              91.1   
Atlantic County Vocational School District             97.7              91.1   
Buena Regional School District                         88.2              91.1   
Egg Harbor Township School District                    93.7              91.1   
Greater Egg Harbor Regional High School District       87.6              91.1   

                                                                      ADDRESS  \
DistrictName                                                                    
Atlantic City School District                            1300 Atlantic Avenue   
Atlantic County Vocational School District               5080 Atlantic Avenue   
Buena Regional School District                                914 Main Avenue   
Egg Harbor Township School District                            13 Swift Drive   
Greater Egg Harbor Regional High School District  1824 Dr. Dennis Foreman Dr.   

                                                                CITY_STATE_ZIP  \
DistrictName                                                                     
Atlantic City School District                           Atlantic City NJ 08401   
Atlantic County Vocational School District               Mays Landing NJ 08330   
Buena Regional School District                               Richland NJ 08350   
Egg Harbor Township School District               Egg Harbor Township NJ 08234   
Greater Egg Harbor Regional High School District    Mays Landing NJ 08330-2640   

                                                                                       full_address  
DistrictName                                                                                         
Atlantic City School District                     1300 Atlantic Avenue, Atlantic City NJ 08401, USA  
Atlantic County Vocational School District         5080 Atlantic Avenue, Mays Landing NJ 08330, USA  
Buena Regional School District                              914 Main Avenue, Richland NJ 08350, USA  
Egg Harbor Township School District               13 Swift Drive, Egg Harbor Township NJ 08234, USA  
Greater Egg Harbor Regional High School District  1824 Dr. Dennis Foreman Dr., Mays Landing NJ 0...

And get some info on it as well.

In [205]:
merged_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 316 entries, Atlantic City School District to University Academy Charter High School
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Graduates         316 non-null    float64
 1   State: Graduates  316 non-null    float64
 2   ADDRESS           316 non-null    object 
 3   CITY_STATE_ZIP    316 non-null    object 
 4   full_address      316 non-null    object 
dtypes: float64(2), object(3)
memory usage: 14.8+ KB


Okay great, there are 316 school districts accounted for! Let's continue and add longitude and latitude coordinates for each given address. To do this, we will create an API key on Google Cloud to be able to use the Google Maps platform.

Let's start by importing the necessary libraries

In [206]:
import requests
import time

And now we can get into it by inserting the API key we created.

In [207]:
API_KEY = 'AIzaSyCLhXMlPpoFK5W2mY5dVW8n95NxkvdkTxo'

And now we can go through each address and find the longitude and latitude coordinates, adding a delay to ensure we do not hit rate limits.

In [208]:
latitudes = []
longitudes = []
for address in merged_data['full_address']:
    url = f"https://maps.googleapis.com/maps/api/geocode/json?address={address}&key={API_KEY}"
    try:
        response = requests.get(url)
        response.raise_for_status()
        data = response.json()
        if data['status'] == 'OK':
            location = data['results'][0]['geometry']['location']
            latitudes.append(location['lat'])
            longitudes.append(location['lng'])
        else:
            print(f"Geocoding error for '{address}': {data['status']}")
            latitudes.append(None)
            longitudes.append(None)
        time.sleep(1)
    except Exception as e:
        print(f"Error with address '{address}': {e}")
        latitudes.append(None)
        longitudes.append(None)

Okay, now we can add our newly found coordinates to our merged dataframe as new columns and check out our results!

In [210]:
merged_data['latitude'] = latitudes
merged_data['longitude'] = longitudes
print(merged_data.head())

                                                  Graduates  State: Graduates  \
DistrictName                                                                    
Atlantic City School District                          79.8              91.1   
Atlantic County Vocational School District             97.7              91.1   
Buena Regional School District                         88.2              91.1   
Egg Harbor Township School District                    93.7              91.1   
Greater Egg Harbor Regional High School District       87.6              91.1   

                                                                      ADDRESS  \
DistrictName                                                                    
Atlantic City School District                            1300 Atlantic Avenue   
Atlantic County Vocational School District               5080 Atlantic Avenue   
Buena Regional School District                                914 Main Avenue   
Egg Harbor Township School 

Perfect! Let's see the info on the dataframe as well.

In [212]:
merged_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 316 entries, Atlantic City School District to University Academy Charter High School
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Graduates         316 non-null    float64
 1   State: Graduates  316 non-null    float64
 2   ADDRESS           316 non-null    object 
 3   CITY_STATE_ZIP    316 non-null    object 
 4   full_address      316 non-null    object 
 5   latitude          316 non-null    float64
 6   longitude         316 non-null    float64
dtypes: float64(4), object(3)
memory usage: 19.8+ KB


All 316 districts accounted for! Now we can proceed.

I would like to display this data on a geographical map with a heatmap overlay, so let's install and import the libraries we will use.

In [213]:
pip install folium

In [214]:
import folium
from folium.plugins import HeatMap
from IPython.display import display

We can start by initiating a base map that is centered on the average location. In this case, all locations are in the state of New Jersey.

In [215]:
map_center = [merged_data['latitude'].mean(), merged_data['longitude'].mean()]
m = folium.Map(location=map_center, zoom_start=10)

We can get the state average graduation rate from the 'State: Graduates' column. In this column, all values are the same, so we can just take the first value.

In [216]:
state_average_grad_rate = merged_data['State: Graduates'].iloc[0]

Now it's time for us to prepare our data for the heatmap! We can do this by retrieving the latitude, longitude, and graduation rate for each district and adding them to the 'heat_data' list that we will initialize. Before being added to the list, the graduation rates will be normalized relative to the state average.

In [217]:
heat_data = []
for _, row in merged_data.iterrows():
    lat = row['latitude']
    lon = row['longitude']
    graduation_rate = float(row['Graduates'])
    normalized_grad_rate = graduation_rate - state_average_grad_rate
    heat_data.append([lat, lon, normalized_grad_rate])

Now that we have all of the relevant data within the list, we can add this heatmap layer to the map. We can also use a custom gradient to make the heatmap more visually appealing.

In [218]:
HeatMap(heat_data, min_opacity=0.5, radius=15, blur=10).add_to(m)

Now let's save our map as an HTML file in case we would like to view it in a browser separately from this notebook and then check it out!

In [219]:
m.save('school_district_grad_heatmap.html')
display(m)

Now that we have explored our target variable, let's go into the factors we will be testing!

I have separated these factors into 3 groups:
- Family-Level
- School-Level, and
- Student-Level

Without further ado, let's get into it!

# Family-Level Factors

We will focus on 4 columns within the `EnrollmentTrendsByStudentGroup` sheet:
- 'Homeless Students',
- 'Students in Foster Care',
- Military-Connected Students, and
- Economically Disadvantaged Students.

First, let's create a dataframe of the sheet, bringing in only our desired factors.

In [220]:
family_factors_df = pd.DataFrame(desired_data['EnrollmentTrendsByStudentGroup'])
family_factors_df = family_factors_df[['DistrictName', 'Homeless Students', 'Students in Foster Care', 'Military-Connected Students', 'Economically Disadvantaged Students']]

Set 'DistrictName' as the index

In [221]:
family_factors_df.set_index('DistrictName', inplace=True)

And let's check it out

In [222]:
family_factors_df.head()

Homeless Students  \
DistrictName                                                          
Absecon Public Schools District                                 1.9   
Atlantic City School District                                   2.0   
Atlantic County Vocational School District                      0.5   
Atlantic County Special Services School District                2.5   
Brigantine Public School District                               1.8   

                                                  Students in Foster Care  \
DistrictName                                                                
Absecon Public Schools District                                       0.4   
Atlantic City School District                                         0.2   
Atlantic County Vocational School District                            0.0   
Atlantic County Special Services School District                      1.5   
Brigantine Public School District                                     0.3   

                                                  Military-Connected Students  \
DistrictName                                                                    
Absecon Public Schools District                                           0.5   
Atlantic City School District                                             0.5   
Atlantic County Vocational School District                                0.9   
Atlantic County Special Services School District                          0.3   
Brigantine Public School District                                         1.6   

                                                  Economically Disadvantaged Students  
DistrictName                                                                           
Absecon Public Schools District                                                  49.4  
Atlantic City School District                                                    87.1  
Atlantic County Vocational School District                                       53.8  
Atlantic County Special Services School District                                 41.1  
Brigantine Public School District                                                31.5

In [223]:
family_factors_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 668 entries, Absecon Public Schools District to State
Data columns (total 4 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Homeless Students                    668 non-null    float64
 1   Students in Foster Care              668 non-null    float64
 2   Military-Connected Students          668 non-null    float64
 3   Economically Disadvantaged Students  668 non-null    float64
dtypes: float64(4)
memory usage: 26.1+ KB


In [224]:
family_factors_df.describe()

Homeless Students  Students in Foster Care  \
count         668.000000               668.000000   
mean            0.828593                 0.151946   
std             1.433234                 0.261891   
min             0.000000                 0.000000   
25%             0.000000                 0.000000   
50%             0.300000                 0.000000   
75%             1.000000                 0.200000   
max            14.600000                 2.100000   

       Military-Connected Students  Economically Disadvantaged Students  
count                   668.000000                           668.000000  
mean                      0.819012                            31.115120  
std                       3.943611                            26.074879  
min                       0.000000                             0.000000  
25%                       0.000000                            10.000000  
50%                       0.200000                            24.000000  
75%                       0.600000                            49.325000  
max                      71.300000                            97.700000

Import necessary libraries

In [225]:
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.metrics import mean_squared_error, r2_score

Now we can bring the graduation rate column from our other dataframe into our `family_factors` dataframe, and drop any rows that have NaNs.

In [226]:
family_factors_df['Graduation Rate'] = grad_data['Graduates']
family_factors_df.dropna(inplace=True)

Let's check the shape of the dataframe after these changes.

In [227]:
print(family_factors_df.shape)

(316, 5)


316 districts, great! And now it's time to split the data into training and validation sets! Here, X represents the features and y represents the target variable. We will preform an 80/20 train/test split.

In [228]:
X = family_factors_df[['Homeless Students', 'Students in Foster Care', 'Military-Connected Students', 'Economically Disadvantaged Students']]
y = family_factors_df['Graduation Rate']
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

Okay great, these factors are ready for use in our models. We are going to create two different types of regression models - Random Forest and XGBoost.

First we create and train a Random Forest model.

In [229]:
rf_model = RandomForestRegressor()
rf_model.fit(X_train, y_train)

RandomForestRegressor()

And then we create and train an XGBoost model.

In [230]:
xgb_model = GradientBoostingRegressor()
xgb_model.fit(X_train, y_train)

GradientBoostingRegressor()

Now we can use GridSearchCV to tune the hyperparameters and find the optimal ones for each model.

To get the best models, we have to first define the model's parameter grid and then perform Grid Search with cross-validation and a regression scoring metric.

We will first do this for Random Forest and then for XGBoost.

In [231]:
param_grid_rf = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 5, 10],
}
grid_search_rf = GridSearchCV(rf_model, param_grid_rf, cv=5, scoring='neg_mean_squared_error')
grid_search_rf.fit(X_train, y_train)
best_rf_model = grid_search_rf.best_estimator_

In [232]:
param_grid_xgb = {
    'n_estimators': [100, 200, 300],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 5, 7],
}
grid_search_xgb = GridSearchCV(xgb_model, param_grid_xgb, cv=5, scoring='neg_mean_squared_error')
grid_search_xgb.fit(X_train, y_train)
best_xgb_model = grid_search_xgb.best_estimator_

Great! Now that we have the best models, let's evaluate their performance and make predictions on the validation set using the trained models.

In [233]:
rf_predictions = best_rf_model.predict(X_val)
xgb_predictions = best_xgb_model.predict(X_val)

With that done, we can calculate the regression metrics such as MSE and R2 for each model and print our results.

In [234]:
rf_mse = mean_squared_error(y_val, rf_predictions)
rf_r2 = r2_score(y_val, rf_predictions)

xgb_mse = mean_squared_error(y_val, xgb_predictions)
xgb_r2 = r2_score(y_val, xgb_predictions)

In [235]:
print(f"Random Forest Regressor:")
print(f"  Mean Squared Error (MSE): {rf_mse}")
print(f"  R-squared (R2): {rf_r2}")

print(f"\nXGBoost Regressor:")
print(f"  Mean Squared Error (MSE): {xgb_mse}")
print(f"  R-squared (R2): {xgb_r2}")

Random Forest Regressor:
  Mean Squared Error (MSE): 86.34355420473199
  R-squared (R2): -0.05517985818297988

XGBoost Regressor:
  Mean Squared Error (MSE): 77.43813247941543
  R-squared (R2): 0.05365075134808084


And now to answer our main question - which factors have the most influence on the target variable? There are a couple of different ways we can find out:
- Feature Importance, and
- Permutation Importance.

Our Random Forest and XGBoost models have a built-in feature importance attribute that can be used after training. The importance values represent the relative contribution of each feature to the model's predictions.

The higher the importance value for a feature, the greater its influence on the model's predictions.

Let's create a third column that holds the average importance values across features, and rank them by this column.

In [236]:
rf_importances = best_rf_model.feature_importances_
xgb_importances = best_xgb_model.feature_importances_
feature_names = ['Homeless Students', 'Students in Foster Care', 'Military-Connected Students', 'Economically Disadvantaged Students']
importance_df = pd.DataFrame({'Feature': feature_names, 'RF Importance': rf_importances, 'XGB Importance': xgb_importances})
importance_df['Avg_Importance'] = (importance_df['RF Importance'] + importance_df['XGB Importance']) / 2
importance_df.sort_values(by=['Avg_Importance'], ascending=False, inplace=True)
print(importance_df)

                               Feature  RF Importance  XGB Importance  \
3  Economically Disadvantaged Students       0.566352        0.596604   
0                    Homeless Students       0.272661        0.282537   
1              Students in Foster Care       0.110385        0.119498   
2          Military-Connected Students       0.050602        0.001361   

   Avg_Importance  
3        0.581478  
0        0.277599  
1        0.114942  
2        0.025981  


Permutation importance, on the other hand, can be found for any model, and it measures the decrease in model performance when the values of a single feature are randomly shuffled. This helps understand how much the model relies on that feature for accurate predictions.

Permutation importance is model-agnostic and provides a more general measure of feature importance based on model performance.

Let's do the same thing here that we did with the other method, and merge the importance results of the two models and then create a third column that holds the average importance values across features, and rank them by this column.

To do this, we will first need to import the library.

In [237]:
from sklearn.inspection import permutation_importance

Now we can calculate and print.

In [238]:
result_rf = permutation_importance(best_rf_model, X_val, y_val, n_repeats=10, random_state=42)
rf_importance_df = pd.DataFrame({'Feature': feature_names, 'RF Importance': result_rf.importances_mean})
result_xgb = permutation_importance(best_xgb_model, X_val, y_val, n_repeats=10, random_state=42)
xgb_importance_df = pd.DataFrame({'Feature': feature_names, 'XGB Importance': result_xgb.importances_mean})
importance_df = pd.merge(rf_importance_df, xgb_importance_df, on='Feature')
importance_df['Avg_Importance'] = (importance_df['RF Importance'] + importance_df['XGB Importance']) / 2
importance_df.sort_values(by=['Avg_Importance'], ascending=False, inplace=True)
print(importance_df)

                               Feature  RF Importance  XGB Importance  \
3  Economically Disadvantaged Students       0.123968        0.145927   
2          Military-Connected Students      -0.002282        0.000111   
0                    Homeless Students      -0.112758        0.003057   
1              Students in Foster Care      -0.110190       -0.071148   

   Avg_Importance  
3        0.134947  
2       -0.001086  
0       -0.054850  
1       -0.090669  


Wow, a unanimous decision of the top spot for feature influence!

Economic Disadvantage has the biggest impact on graduation rate in this category.

Let's move on to the the next area of influence category: the school.

# School-Level Factors

Here, I will use the following 4 columns from the following 3 sheets:

▶  `TeachersExperience` : 2 columns

- 'TeacherAvgYearsExp_District'
- 'PercentageOfOutOfFieldTeachers_District'

▶ `StudentToStaffRatios` : 1 column

- 'Student_Teacher_District'

▶ `TeachersAdminsLevelOfEducation` : 1 column

- 'MastersOrHigher' - To be created as a new column within the `TeachersAdminsLevelOfEducation` sheet in the prefered_data by adding the values in the columns 'Masters' and 'Doctoral' for each entry



Let's make a dataframe holding all of these factors, except that for the `TeachersAdminsLevelOfEducation` and `TeachersAdminsOneYearRetention` sheets let's only add an entry to the dataframe if the 'Teachers/Admins' column value for that entry is "Teachers". Also, let's make sure that the values being used for the 'Student_Teacher_District' dataframe have the ':1' removed at the end.

To do this, we will have to process the data from one sheet at a time by creating dataframes for each of them, set the index in each dataframe to the 'DistrictName' column, and then merge the dataframes together based on this index. We can also check out our dataframes as we go along.

Let's start with `TeachersExperience`

In [239]:
teachers_exp_df = desired_data['TeachersExperience'][['DistrictName', 'TeacherAvgYearsExp_District', 'PercentageOfOutOfFieldTeachers_District']]
teachers_exp_df.set_index('DistrictName', inplace=True)

In [240]:
teachers_exp_df.head()

TeacherAvgYearsExp_District  \
DistrictName                                                                    
Absecon Public Schools District                                          12.4   
Atlantic City School District                                            14.2   
Atlantic County Vocational School District                                7.8   
Atlantic County Special Services School District                         16.0   
Brigantine Public School District                                        17.2   

                                                  PercentageOfOutOfFieldTeachers_District  
DistrictName                                                                               
Absecon Public Schools District                                                       1.1  
Atlantic City School District                                                         0.8  
Atlantic County Vocational School District                                            4.9  
Atlantic County Special Services School District                                      0.0  
Brigantine Public School District                                                     0.0

In [241]:
teachers_exp_df.describe()

TeacherAvgYearsExp_District  PercentageOfOutOfFieldTeachers_District
count                   667.000000                               667.000000
mean                     12.001349                                 4.011094
std                       2.849137                                 6.499427
min                       0.900000                                 0.000000
25%                      10.500000                                 0.400000
50%                      12.300000                                 1.800000
75%                      13.950000                                 4.300000
max                      22.400000                                55.000000

Now the same with `StudentToStaffRatios`

In [242]:
students_per_staff_df = desired_data['StudentToStaffRatios'][['DistrictName', 'Student_Teacher_District']]
students_per_staff_df['Student_Teacher_District'] = students_per_staff_df['Student_Teacher_District'].str.replace(':1', '', regex=False)
students_per_staff_df['Student_Teacher_District'] = students_per_staff_df['Student_Teacher_District'].astype(float)
students_per_staff_df.set_index('DistrictName', inplace=True)

<ipython-input-242-af817703207c>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  students_per_staff_df['Student_Teacher_District'] = students_per_staff_df['Student_Teacher_District'].str.replace(':1', '', regex=False)
<ipython-input-242-af817703207c>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  students_per_staff_df['Student_Teacher_District'] = students_per_staff_df['Student_Teacher_District'].astype(float)


In [243]:
students_per_staff_df.head()

Student_Teacher_District
DistrictName                                                              
Absecon Public Schools District                                       10.0
Atlantic City School District                                         10.0
Atlantic County Vocational School District                            12.0
Atlantic County Special Services School District                      13.0
Brigantine Public School District                                      6.0

In [244]:
students_per_staff_df.describe()

Student_Teacher_District
count                668.000000
mean                  11.693114
std                   17.293398
min                    1.000000
25%                    9.000000
50%                   11.000000
75%                   12.000000
max                  449.000000

And now the `TeachersAdminsLevelOfEducation` sheet

In [245]:
desired_data['TeachersAdminsLevelOfEducation']['Masters'] = pd.to_numeric(desired_data['TeachersAdminsLevelOfEducation']['Masters'], errors='coerce')
desired_data['TeachersAdminsLevelOfEducation']['Doctoral'] = pd.to_numeric(desired_data['TeachersAdminsLevelOfEducation']['Doctoral'], errors='coerce')
desired_data['TeachersAdminsLevelOfEducation']['MastersOrHigher'] = desired_data['TeachersAdminsLevelOfEducation']['Masters'] + desired_data['TeachersAdminsLevelOfEducation']['Doctoral']
teachers_edu_level_df = desired_data['TeachersAdminsLevelOfEducation'][desired_data['TeachersAdminsLevelOfEducation']['Teachers/Admins'] == 'Teachers']
teachers_edu_level_df = teachers_edu_level_df[['DistrictName','MastersOrHigher']]
teachers_edu_level_df.set_index('DistrictName', inplace=True)

In [246]:
teachers_edu_level_df.head()

MastersOrHigher
DistrictName                                                     
Absecon Public Schools District                              39.6
Atlantic City School District                                46.8
Atlantic County Vocational School District                   22.2
Atlantic County Special Services School District             32.0
Brigantine Public School District                            41.7

In [247]:
teachers_edu_level_df.describe()

MastersOrHigher
count       665.000000
mean         42.657293
std          16.413573
min           0.000000
25%          31.000000
50%          41.200000
75%          54.200000
max         100.000000

Nice! Now that we have processed all of the individual sheets of factors, we can merge the dataframes based on their common index.

In [255]:
school_factors_merged_df = teachers_exp_df.join(students_per_staff_df, how='inner').join(teachers_edu_level_df, how='inner')

In [256]:
school_factors_merged_df.head()

TeacherAvgYearsExp_District  \
DistrictName                                                                    
Absecon Public Schools District                                          12.4   
Atlantic City School District                                            14.2   
Atlantic County Vocational School District                                7.8   
Atlantic County Special Services School District                         16.0   
Brigantine Public School District                                        17.2   

                                                  PercentageOfOutOfFieldTeachers_District  \
DistrictName                                                                                
Absecon Public Schools District                                                       1.1   
Atlantic City School District                                                         0.8   
Atlantic County Vocational School District                                            4.9   
Atlantic County Special Services School District                                      0.0   
Brigantine Public School District                                                     0.0   

                                                  Student_Teacher_District  \
DistrictName                                                                 
Absecon Public Schools District                                       10.0   
Atlantic City School District                                         10.0   
Atlantic County Vocational School District                            12.0   
Atlantic County Special Services School District                      13.0   
Brigantine Public School District                                      6.0   

                                                  MastersOrHigher  
DistrictName                                                       
Absecon Public Schools District                              39.6  
Atlantic City School District                                46.8  
Atlantic County Vocational School District                   22.2  
Atlantic County Special Services School District             32.0  
Brigantine Public School District                            41.7

Cool! Now that we have the school-based factors ready, let's add the graduation rates back in, being sure to remove any rows with missing values (NaNs).

In [259]:
school_factors_merged_df['Graduation Rate'] = grad_data['Graduates']
school_factors_merged_df.dropna(inplace=True)

In [260]:
school_factors_merged_df.head()

TeacherAvgYearsExp_District  \
DistrictName                                                                    
Atlantic City School District                                            14.2   
Atlantic County Vocational School District                                7.8   
Buena Regional School District                                           14.3   
Egg Harbor Township School District                                      13.6   
Greater Egg Harbor Regional High School District                         14.1   

                                                  PercentageOfOutOfFieldTeachers_District  \
DistrictName                                                                                
Atlantic City School District                                                         0.8   
Atlantic County Vocational School District                                            4.9   
Buena Regional School District                                                        2.8   
Egg Harbor Township School District                                                   3.3   
Greater Egg Harbor Regional High School District                                      0.8   

                                                  Student_Teacher_District  \
DistrictName                                                                 
Atlantic City School District                                         10.0   
Atlantic County Vocational School District                            12.0   
Buena Regional School District                                        11.0   
Egg Harbor Township School District                                   11.0   
Greater Egg Harbor Regional High School District                      11.0   

                                                  MastersOrHigher  \
DistrictName                                                        
Atlantic City School District                                46.8   
Atlantic County Vocational School District                   22.2   
Buena Regional School District                               30.5   
Egg Harbor Township School District                          31.1   
Greater Egg Harbor Regional High School District             41.8   

                                                  Graduation Rate  
DistrictName                                                       
Atlantic City School District                                79.8  
Atlantic County Vocational School District                   97.7  
Buena Regional School District                               88.2  
Egg Harbor Township School District                          93.7  
Greater Egg Harbor Regional High School District             87.6

Okay, now our data is ready to be split. We are going to follow the same method as with the first group of factors, doing an 80/20 train/test split and using X and y to represent the features and target variable, respectively.

In [262]:
X = school_factors_merged_df[['TeacherAvgYearsExp_District', 'PercentageOfOutOfFieldTeachers_District', 'Student_Teacher_District', 'MastersOrHigher']]
y = school_factors_merged_df['Graduation Rate']
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

Okay great, these factors are ready for use in our models.

Again we will create and train a Random Forest and an XGBoost model, but this time trained on our school-level factors as opposed to our family-level factors.

In [263]:
rf_model = RandomForestRegressor()
rf_model.fit(X_train, y_train)

RandomForestRegressor()

In [264]:
xgb_model = GradientBoostingRegressor()
xgb_model.fit(X_train, y_train)

GradientBoostingRegressor()

Just like the first time, we will now tune the hyperparameters for each model until we get the best of each model.

In [265]:
param_grid_rf = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 5, 10],
}
grid_search_rf = GridSearchCV(rf_model, param_grid_rf, cv=5, scoring='neg_mean_squared_error')
grid_search_rf.fit(X_train, y_train)
best_rf_model = grid_search_rf.best_estimator_

In [266]:
param_grid_xgb = {
    'n_estimators': [100, 200, 300],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 5, 7],
}
grid_search_xgb = GridSearchCV(xgb_model, param_grid_xgb, cv=5, scoring='neg_mean_squared_error')
grid_search_xgb.fit(X_train, y_train)
best_xgb_model = grid_search_xgb.best_estimator_

Now we can evaluate the new models' prediction abilities with the test set.

In [267]:
rf_predictions = best_rf_model.predict(X_val)
xgb_predictions = best_xgb_model.predict(X_val)

Let's calculate and print the results!

In [268]:
rf_mse = mean_squared_error(y_val, rf_predictions)
rf_r2 = r2_score(y_val, rf_predictions)
xgb_mse = mean_squared_error(y_val, xgb_predictions)
xgb_r2 = r2_score(y_val, xgb_predictions)

In [269]:
print(f"Random Forest Regressor:")
print(f"  Mean Squared Error (MSE): {rf_mse}")
print(f"  R-squared (R2): {rf_r2}")
print(f"\nXGBoost Regressor:")
print(f"  Mean Squared Error (MSE): {xgb_mse}")
print(f"  R-squared (R2): {xgb_r2}")

Random Forest Regressor:
  Mean Squared Error (MSE): 37.230878935662815
  R-squared (R2): -0.6807715833604895

XGBoost Regressor:
  Mean Squared Error (MSE): 29.463463798792585
  R-squared (R2): -0.3301150581472152


In terms of the influence the individual features had on the target variable, we can once again start by looking at the built-in feature importance attribute of our models and then move on to using permutation importance. Let's also create the third average importance column again, and sort the features by it.

In [270]:
rf_importances = best_rf_model.feature_importances_
xgb_importances = best_xgb_model.feature_importances_
feature_names = ['TeacherAvgYearsExp_District', 'PercentageOfOutOfFieldTeachers_District', 'Student_Teacher_District', 'MastersOrHigher']
importance_df = pd.DataFrame({'Feature': feature_names, 'RF Importance': rf_importances, 'XGB Importance': xgb_importances})
importance_df['Avg_Importance'] = (importance_df['RF Importance'] + importance_df['XGB Importance']) / 2
importance_df.sort_values(by=['Avg_Importance'], ascending=False, inplace=True)
print(importance_df)

                                   Feature  RF Importance  XGB Importance  \
2                 Student_Teacher_District       0.401448        0.647256   
3                          MastersOrHigher       0.302719        0.194491   
1  PercentageOfOutOfFieldTeachers_District       0.171374        0.058075   
0              TeacherAvgYearsExp_District       0.124459        0.100178   

   Avg_Importance  
2        0.524352  
3        0.248605  
1        0.114725  
0        0.112319  


In [271]:
result_rf = permutation_importance(best_rf_model, X_val, y_val, n_repeats=10, random_state=42)
rf_importance_df = pd.DataFrame({'Feature': feature_names, 'RF Importance': result_rf.importances_mean})
result_xgb = permutation_importance(best_xgb_model, X_val, y_val, n_repeats=10, random_state=42)
xgb_importance_df = pd.DataFrame({'Feature': feature_names, 'XGB Importance': result_xgb.importances_mean})
importance_df = pd.merge(rf_importance_df, xgb_importance_df, on='Feature')
importance_df['Avg_Importance'] = (importance_df['RF Importance'] + importance_df['XGB Importance']) / 2
importance_df.sort_values(by=['Avg_Importance'], ascending=False, inplace=True)
print(importance_df)

                                   Feature  RF Importance  XGB Importance  \
2                 Student_Teacher_District       0.417246        0.255284   
0              TeacherAvgYearsExp_District       0.341910        0.163052   
3                          MastersOrHigher       0.232399        0.136892   
1  PercentageOfOutOfFieldTeachers_District      -0.014614       -0.002392   

   Avg_Importance  
2        0.336265  
0        0.252481  
3        0.184645  
1       -0.008503  


Once again, a unanimous decision for the spot of top influence for this category!

Student-Teacher Ratios have the biggest impact on the target variable here.

Let's move on to the final level of factors: the students themselves.

# Student-Level Factors

Here, I will use the following 7 columns from the following 4 sheets:

▶  `PSAT-SAT-ACTParticipation` : 3 columns

- 'SAT'
- 'ACT'
- 'PSAT'

▶ `APIBCourseworkPartPerf` : 2 columns

- 'APIB_COURSE_DISTRICT'
- 'APIB_EXAM_DISTRICT'

▶ `ChronicAbsenteeism` : 1 column

- 'Chronic_Abs_Pct'

▶ `ViolenceVandalismHIBSubstanceOf` : 1 column

- 'Incidents Per 100 Students Enrolled'



Let's make a dataframe holding all of these factors, with the exception of the `ChronicAbsenteeism` sheet where we will only add an entry to the dataframe if the 'StudentGroup' column value for that entry is "Districtwide". Then let's set the index to 'DistrictName', after changing all factors to floats if they are not already.

First let's process data from the `PSAT-SAT-ACTParticipation` sheet

In [276]:
psat_sat_act_df = desired_data['PSAT-SAT-ACTParticipation'][['DistrictName','SAT', 'ACT', 'PSAT']]
psat_sat_act_df['SAT'] = pd.to_numeric(psat_sat_act_df['SAT'], errors='coerce').astype(float)
psat_sat_act_df['ACT'] = pd.to_numeric(psat_sat_act_df['ACT'], errors='coerce').astype(float)
psat_sat_act_df.set_index('DistrictName', inplace=True)

<ipython-input-276-e8743b9985ec>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  psat_sat_act_df['SAT'] = pd.to_numeric(psat_sat_act_df['SAT'], errors='coerce').astype(float)
<ipython-input-276-e8743b9985ec>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  psat_sat_act_df['ACT'] = pd.to_numeric(psat_sat_act_df['ACT'], errors='coerce').astype(float)


In [277]:
psat_sat_act_df.head()

SAT  ACT  PSAT
DistrictName                                                     
Atlantic City School District                     36.9  3.2  74.9
Atlantic County Vocational School District        48.7  0.0  95.7
Buena Regional School District                    28.8  0.9  64.9
Egg Harbor Township School District               49.8  0.9  90.8
Greater Egg Harbor Regional High School District  89.5  2.4  89.1

In [279]:
psat_sat_act_df.describe()

SAT         ACT        PSAT
count  314.000000  314.000000  319.000000
mean    61.664968    8.537261   78.372100
std     21.771715   13.398764   24.825171
min      1.300000    0.000000    0.000000
25%     46.075000    1.100000   64.750000
50%     62.050000    3.300000   90.800000
75%     75.825000   10.750000   95.050000
max    100.000000  100.000000  100.000000

Now the `APIBCourseworkPartPerf` sheet

In [280]:
ap_ib_df = desired_data['APIBCourseworkPartPerf'][['DistrictName','APIB_COURSE_DISTRICT', 'APIB_EXAM_DISTRICT']]
ap_ib_df.set_index('DistrictName', inplace=True)

In [281]:
ap_ib_df.head()

APIB_COURSE_DISTRICT  \
DistrictName                                                             
Atlantic City School District                                     20.7   
Atlantic County Vocational School District                        15.2   
Buena Regional School District                                    27.5   
Egg Harbor Township School District                               32.8   
Greater Egg Harbor Regional High School District                  25.6   

                                                  APIB_EXAM_DISTRICT  
DistrictName                                                          
Atlantic City School District                                   15.1  
Atlantic County Vocational School District                      10.0  
Buena Regional School District                                  10.2  
Egg Harbor Township School District                             29.3  
Greater Egg Harbor Regional High School District                14.6

In [282]:
ap_ib_df.describe()

APIB_COURSE_DISTRICT  APIB_EXAM_DISTRICT
count            314.000000          314.000000
mean              34.550955           30.044586
std               19.819944           17.934289
min                0.000000            0.000000
25%               20.375000           17.150000
50%               31.100000           27.100000
75%               48.900000           43.075000
max               99.500000           91.900000

And the `ChronicAbsenteeism` sheet

In [286]:
chronic_abs_df = desired_data['ChronicAbsenteeism'][desired_data['ChronicAbsenteeism']['StudentGroup'] == 'Districtwide'][['DistrictName','Chronic_Abs_Pct']]
chronic_abs_df['Chronic_Abs_Pct'] = pd.to_numeric(chronic_abs_df['Chronic_Abs_Pct'], errors='coerce').astype(float)
chronic_abs_df.set_index('DistrictName', inplace=True)

In [287]:
chronic_abs_df.head()

Chronic_Abs_Pct
DistrictName                                               
Absecon Public Schools District                        16.3
Atlantic City School District                          31.3
Atlantic County Vocational School District             18.3
Brigantine Public School District                      24.8
Buena Regional School District                         23.5

In [288]:
chronic_abs_df.describe()

Chronic_Abs_Pct
count       647.000000
mean         15.064606
std           7.664067
min           0.000000
25%           9.500000
50%          13.800000
75%          19.100000
max          50.800000

And finally the `ViolenceVandalismHIBSubstanceOf` sheet.

In [290]:
incidents_df = desired_data['ViolenceVandalismHIBSubstanceOf'][['DistrictName','Incidents Per 100 Students Enrolled']]
incidents_df.set_index('DistrictName', inplace=True)

In [291]:
incidents_df.head()

Incidents Per 100 Students Enrolled
DistrictName                                                                         
Absecon Public Schools District                                                  0.84
Atlantic City School District                                                    3.45
Atlantic County Vocational School District                                       2.15
Atlantic County Special Services School District                                 5.21
Brigantine Public School District                                                1.82

In [292]:
incidents_df.describe()

Incidents Per 100 Students Enrolled
count                           668.000000
mean                              2.604805
std                               3.829248
min                               0.000000
25%                               0.797500
50%                               1.705000
75%                               3.217500
max                              59.260000

We now have all of our student-level data!

Let's merge these dataframes using their common index of 'DistrictName'.

In [303]:
school_factors_merged_df = teachers_exp_df.join(students_per_staff_df, how='inner').join(teachers_edu_level_df, how='inner')

In [304]:
student_factors_merged_df.head()

SAT  ACT  PSAT  \
DistrictName                                                        
Atlantic City School District                     36.9  3.2  74.9   
Atlantic County Vocational School District        48.7  0.0  95.7   
Buena Regional School District                    28.8  0.9  64.9   
Egg Harbor Township School District               49.8  0.9  90.8   
Greater Egg Harbor Regional High School District  89.5  2.4  89.1   

                                                  APIB_COURSE_DISTRICT  \
DistrictName                                                             
Atlantic City School District                                     20.7   
Atlantic County Vocational School District                        15.2   
Buena Regional School District                                    27.5   
Egg Harbor Township School District                               32.8   
Greater Egg Harbor Regional High School District                  25.6   

                                                  APIB_EXAM_DISTRICT  \
DistrictName                                                           
Atlantic City School District                                   15.1   
Atlantic County Vocational School District                      10.0   
Buena Regional School District                                  10.2   
Egg Harbor Township School District                             29.3   
Greater Egg Harbor Regional High School District                14.6   

                                                  Chronic_Abs_Pct  \
DistrictName                                                        
Atlantic City School District                                31.3   
Atlantic County Vocational School District                   18.3   
Buena Regional School District                               23.5   
Egg Harbor Township School District                          18.1   
Greater Egg Harbor Regional High School District             24.4   

                                                  Incidents Per 100 Students Enrolled  \
DistrictName                                                                            
Atlantic City School District                                                    3.45   
Atlantic County Vocational School District                                       2.15   
Buena Regional School District                                                   4.22   
Egg Harbor Township School District                                              2.40   
Greater Egg Harbor Regional High School District                                 8.92   

                                                  Graduation Rate  
DistrictName                                                       
Atlantic City School District                                79.8  
Atlantic County Vocational School District                   97.7  
Buena Regional School District                               88.2  
Egg Harbor Township School District                          93.7  
Greater Egg Harbor Regional High School District             87.6

And now we can add our graduation rates back in the same way, once again being sure to remove any data points with NaNs.

In [308]:
student_factors_merged_df['Graduation Rate'] = grad_data['Graduates']
student_factors_merged_df.dropna(inplace=True)

In [312]:
student_factors_merged_df.head()

SAT  ACT  PSAT  \
DistrictName                                                        
Atlantic City School District                     36.9  3.2  74.9   
Atlantic County Vocational School District        48.7  0.0  95.7   
Buena Regional School District                    28.8  0.9  64.9   
Egg Harbor Township School District               49.8  0.9  90.8   
Greater Egg Harbor Regional High School District  89.5  2.4  89.1   

                                                  APIB_COURSE_DISTRICT  \
DistrictName                                                             
Atlantic City School District                                     20.7   
Atlantic County Vocational School District                        15.2   
Buena Regional School District                                    27.5   
Egg Harbor Township School District                               32.8   
Greater Egg Harbor Regional High School District                  25.6   

                                                  APIB_EXAM_DISTRICT  \
DistrictName                                                           
Atlantic City School District                                   15.1   
Atlantic County Vocational School District                      10.0   
Buena Regional School District                                  10.2   
Egg Harbor Township School District                             29.3   
Greater Egg Harbor Regional High School District                14.6   

                                                  Chronic_Abs_Pct  \
DistrictName                                                        
Atlantic City School District                                31.3   
Atlantic County Vocational School District                   18.3   
Buena Regional School District                               23.5   
Egg Harbor Township School District                          18.1   
Greater Egg Harbor Regional High School District             24.4   

                                                  Incidents Per 100 Students Enrolled  \
DistrictName                                                                            
Atlantic City School District                                                    3.45   
Atlantic County Vocational School District                                       2.15   
Buena Regional School District                                                   4.22   
Egg Harbor Township School District                                              2.40   
Greater Egg Harbor Regional High School District                                 8.92   

                                                  Graduation Rate  
DistrictName                                                       
Atlantic City School District                                79.8  
Atlantic County Vocational School District                   97.7  
Buena Regional School District                               88.2  
Egg Harbor Township School District                          93.7  
Greater Egg Harbor Regional High School District             87.6

Now for the final time we can split the data into training and test sets, in the same manner that we did in the other two levels.

In [314]:
X = student_factors_merged_df[['SAT', 'ACT', 'PSAT', 'APIB_COURSE_DISTRICT', 'APIB_EXAM_DISTRICT', 'Chronic_Abs_Pct', 'Incidents Per 100 Students Enrolled']]
y = student_factors_merged_df['Graduation Rate']
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

Now we can create our two models based on the new training set...

In [315]:
rf_model = RandomForestRegressor()
rf_model.fit(X_train, y_train)

RandomForestRegressor()

In [316]:
xgb_model = GradientBoostingRegressor()
xgb_model.fit(X_train, y_train)

GradientBoostingRegressor()

...before tuning the hyperparameters for each model.

In [317]:
param_grid_rf = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 5, 10],
}
grid_search_rf = GridSearchCV(rf_model, param_grid_rf, cv=5, scoring='neg_mean_squared_error')
grid_search_rf.fit(X_train, y_train)
best_rf_model = grid_search_rf.best_estimator_

In [318]:
param_grid_xgb = {
    'n_estimators': [100, 200, 300],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 5, 7],
}
grid_search_xgb = GridSearchCV(xgb_model, param_grid_xgb, cv=5, scoring='neg_mean_squared_error')
grid_search_xgb.fit(X_train, y_train)
best_xgb_model = grid_search_xgb.best_estimator_

Now we can make our predictions with the trained models...

In [319]:
rf_predictions = best_rf_model.predict(X_val)
xgb_predictions = best_xgb_model.predict(X_val)

...before calculating and printing the results.

In [320]:
rf_mse = mean_squared_error(y_val, rf_predictions)
rf_r2 = r2_score(y_val, rf_predictions)
xgb_mse = mean_squared_error(y_val, xgb_predictions)
xgb_r2 = r2_score(y_val, xgb_predictions)

In [321]:
print(f"Random Forest Regressor:")
print(f"  Mean Squared Error (MSE): {rf_mse}")
print(f"  R-squared (R2): {rf_r2}")
print(f"\nXGBoost Regressor:")
print(f"  Mean Squared Error (MSE): {xgb_mse}")
print(f"  R-squared (R2): {xgb_r2}")

Random Forest Regressor:
  Mean Squared Error (MSE): 45.297531896853116
  R-squared (R2): 0.4053605329705203

XGBoost Regressor:
  Mean Squared Error (MSE): 52.285183786300095
  R-squared (R2): 0.31363073177980794


To analyze the influence these final individual features had on the target variable, we can again look at the feature importance and permutation importance of each model. Let's also create the third average importance column again, and sort the features by it.

In [322]:
rf_importances = best_rf_model.feature_importances_
xgb_importances = best_xgb_model.feature_importances_
feature_names = ['SAT', 'ACT', 'PSAT', 'APIB_COURSE_DISTRICT', 'APIB_EXAM_DISTRICT', 'Chronic_Abs_Pct', 'Incidents Per 100 Students Enrolled']
importance_df = pd.DataFrame({'Feature': feature_names, 'RF Importance': rf_importances, 'XGB Importance': xgb_importances})
importance_df['Avg_Importance'] = (importance_df['RF Importance'] + importance_df['XGB Importance']) / 2
importance_df.sort_values(by=['Avg_Importance'], ascending=False, inplace=True)
print(importance_df)

                               Feature  RF Importance  XGB Importance  \
5                      Chronic_Abs_Pct       0.521055        0.445445   
3                 APIB_COURSE_DISTRICT       0.133145        0.155569   
4                   APIB_EXAM_DISTRICT       0.121445        0.165217   
0                                  SAT       0.082211        0.108088   
6  Incidents Per 100 Students Enrolled       0.054551        0.066424   
2                                 PSAT       0.049242        0.049312   
1                                  ACT       0.038351        0.009945   

   Avg_Importance  
5        0.483250  
3        0.144357  
4        0.143331  
0        0.095149  
6        0.060488  
2        0.049277  
1        0.024148  


In [323]:
result_rf = permutation_importance(best_rf_model, X_val, y_val, n_repeats=10, random_state=42)
rf_importance_df = pd.DataFrame({'Feature': feature_names, 'RF Importance': result_rf.importances_mean})
result_xgb = permutation_importance(best_xgb_model, X_val, y_val, n_repeats=10, random_state=42)
xgb_importance_df = pd.DataFrame({'Feature': feature_names, 'XGB Importance': result_xgb.importances_mean})
importance_df = pd.merge(rf_importance_df, xgb_importance_df, on='Feature')
importance_df['Avg_Importance'] = (importance_df['RF Importance'] + importance_df['XGB Importance']) / 2
importance_df.sort_values(by=['Avg_Importance'], ascending=False, inplace=True)
print(importance_df)

                               Feature  RF Importance  XGB Importance  \
5                      Chronic_Abs_Pct       0.400994        0.562435   
3                 APIB_COURSE_DISTRICT       0.106789        0.248199   
4                   APIB_EXAM_DISTRICT       0.036505        0.069637   
6  Incidents Per 100 Students Enrolled       0.032903       -0.036357   
2                                 PSAT       0.028667       -0.056158   
1                                  ACT      -0.018683       -0.024550   
0                                  SAT      -0.031419       -0.067705   

   Avg_Importance  
5        0.481715  
3        0.177494  
4        0.053071  
6       -0.001727  
2       -0.013745  
1       -0.021616  
0       -0.049562  


Even more interesting - the top 3 spots of features listed by both average importance methods are unanimous!

For this group of factors, Chronic Absenteeism, AP/IB Course Registration, and AP/IB Exam Registration respectively had the biggest impact on the target variable!

This concludes my second draft.

---------------------------------

**Note to Reader (AKA Professor Gutu):**

My project is in accordance with the objective timeline as listed on my project overview statement, but I realize it is a late submission. I have had a rough week and on top of it I put way too much pressure on myself to submit the best work I can right now as opposed to just submitting it on time with more mediocre work. I understand there are consequences to this.

Regardless, I am very much looking forward to hearing your comments/critiques for this draft! My hope is to be able to provide a great final project, so I appreciate all the feedback I can get!

Thank you